website/project/model.py

```diff
@@ -2825,149 +2863,190 @@ def to_json(self):
-            for node in parent_registration.get_descendants_recursive():
-                node.update_search()
-
-
-def validate_embargo_state(value):
-    acceptable_states = [
-        Embargo.UNAPPROVED, Embargo.ACTIVE, Embargo.CANCELLED, Embargo.COMPLETED
-    ]
-    if value not in acceptable_states:
-        raise ValidationValueError('Invalid embargo state assignment.')
-    return True
+    def reject(self, user, token):
+        """Cancels sanction if user is admin and token verifies."""
+        try:
+            if self.approval_state[user._id]['disapproval_token'] != token:
+                raise self.RejectionInvalidToken
+        except KeyError:
+            raise self.RejectionNotAuthorized
+        self.state = Sanction.CANCELLED
+        self._on_reject(user, token)
 
+    def _notify_authorizer(self, user):
+        pass
 
-class Embargo(StoredObject):
-    """Embargo object for registrations waiting to go public."""
+    def _notify_non_authorizer(self, user):
+        pass
 
-    UNAPPROVED = 'unapproved'
-    ACTIVE = 'active'
-    CANCELLED = 'cancelled'
-    COMPLETED = 'completed'
+    def ask(self, group):
+        for contrib in group:
+            if contrib._id in self.approval_state:
+                self._notify_authorizer(contrib)
+            else:
+                self._notify_non_authorizer(contrib)
 
-    _id = fields.StringField(primary=True, default=lambda: str(ObjectId()))
-    initiation_date = fields.DateTimeField(auto_now_add=datetime.datetime.utcnow)
-    initiated_by = fields.ForeignField('user', backref='embargoed')
-    end_date = fields.DateTimeField()
-    # Expanded: Dictionary field mapping admin IDs their approval status and relevant tokens:
+class EmailApprovableSanction(Sanction):
+
+    AUTHORIZER_NOTIFY_TEMPLATE = None
+    NON_AUTHORIZER_NOTIFY_TEMPLATE = None
+
+    # Store a persistant copy of urls for use when needed outside of a request context.
+    # This field gets automagically updated whenever models approval_state is modified
+    # and the model is saved
     # {
-    #   'b3k97': {
-    #     'has_approved': False,
-    #     'approval_token': 'Pew7wj1Puf7DENUPFPnXSwa1rf3xPN',
-    #     'disapproval_token': 'TwozClTFOic2PYxHDStby94bCQMwJy'}
+    #   'abcde': {
+    #     'approve': [APPROVAL_URL],
+    #     'reject': [REJECT_URL],
+    #   }
     # }
-    approval_state = fields.DictionaryField()
-    # One of 'unapproved', 'active', 'cancelled', or 'completed
-    state = fields.StringField(default='unapproved', validate=validate_embargo_state)
+    stashed_urls = fields.DictionaryField(default={})
+
+    def _view_url(self, user_id):
+        return ''
+
+    def _approval_url(self, user_id):
+        return ''
+
+    def _rejection_url(self, user_id):
+        return ''
+
+    def _send_approval_request_email(self, user, template, context):
+        mails.send_mail(
+            self.initiated_by,
+            template,
+            user=user,
+            **context
```

Comment:
This might break how the email templates currently work. Once we get tests passing we need to verify.